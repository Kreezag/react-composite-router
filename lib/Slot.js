/* eslint-env browser */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import connectRouter from './connect';
import { router as routerParams } from './propTypes';
import { assertTrimmedNonEmptyString, assertPlainObject } from './assert';




class Slot extends Component {
  static contextTypes = {
    router: routerParams.isRequired
  };

  static propTypes = {
    name: PropTypes.string.isRequired,
    props: PropTypes.object,
    render: PropTypes.func,
    children: PropTypes.oneOfType([ PropTypes.func, PropTypes.node ])
  };

  render () {
    const { name, children, render, props = {} } = this.props;
    const { router } = this.context;

    if (process.env.NODE_ENV !== 'production') {
      assertTrimmedNonEmptyString('Slot name', name);
      assertPlainObject('Slot props', props);
    }

    const currentState = router.getState();

    if (currentState.composition.hasOwnProperty(name)) {
      const ViewComponent = currentState.composition[name];

      if (!render) {
        return (
          <ViewComponent {...props} />
        );
      }

      return render(ViewComponent, currentState.params, currentState.name);
    }

    if (!children) {
      return null;
    }

    if (typeof children === 'function') {
      return children(currentState.params, currentState.name);
    }

    return children;
  }
}



export default connectRouter()(Slot);
