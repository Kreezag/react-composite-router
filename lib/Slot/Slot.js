/* eslint-env browser */

import { PropTypes } from 'react';
import { compose, getContext } from 'recompose';
import { connect } from 'react-redux';
import SlotComponent from './Slot.jsx';



export default compose(
  getContext({
    routerSlotsComposition: PropTypes.objectOf(PropTypes.func.isRequired)
  }),
  connect(
    ({ routerState }, { routerSlotsComposition }) => {
      console.log(1111, routerState.name, Object.keys(routerSlotsComposition));

      return ({
        routerState,
        routerSlotsComposition
      });
    }
  )
)(SlotComponent);
