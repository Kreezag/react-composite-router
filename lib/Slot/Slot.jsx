/* eslint-env browser */

import React, { PropTypes } from 'react';
import isFunction from 'lodash/isFunction';
import { assertTrimmedNonEmptyString, assertPlainObject } from '../util/assert';



const Slot = (componentProps, context) => {
  const { name, children, render, props = {}, routerRoutesByName, routerState } = componentProps;

  if (process.env.NODE_ENV !== 'production') {
    assertTrimmedNonEmptyString('Slot name', name);
    assertPlainObject('Slot props', props);
  }

  const routerSlotsComposition = routerRoutesByName.hasOwnProperty(routerState.name) ? routerRoutesByName[routerState.name].slots : {}

  if (routerSlotsComposition.hasOwnProperty(name)) {
    const ViewComponent = routerSlotsComposition[name];

    if (!render) {
      return (<div><ViewComponent/><span>{Object.keys(routerSlotsComposition).join(',')}</span></div>);
    }

    return render(ViewComponent, routerState.params, routerState.name);
  }

  if (!children) {
    return null;
  }

  if (isFunction(children)) {
    return children(routerState.params, routerState.name);
  }

  return children;
};



Slot.propTypes = {
  name: PropTypes.string.isRequired,
  props: PropTypes.object,
  render: PropTypes.func,
  children: PropTypes.oneOfType([ PropTypes.func, PropTypes.element ]),
  routerState: PropTypes.shape({ name: PropTypes.string.isRequired, params: PropTypes.object.isRequired })
};



Slot.defaultProps = {
};



export default Slot;
