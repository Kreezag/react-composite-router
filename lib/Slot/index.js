/* eslint-env browser */

import Slot from './Slot';



export default Slot;
