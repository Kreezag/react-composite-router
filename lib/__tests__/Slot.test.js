/* eslint-env browser, jest */
/* eslint-disable */

import React from 'react';
import Slot from '../Slot';
import renderer from 'react-test-renderer';
import WithContext from 'react-with-context';
import store from './__mocks__/redux-store';

describe('rendering', () => {
  it('should render something', () => {
    const component = renderer.create(
      <WithContext context={{
        store: store({ routerState: { name: '', params: {} } }),
        router: {
          moveTo: () => {},
          onStateChange: () => {},
          getState: () => ({ name: '', params: {}, composition: {} }),
          createHref: () => ({ pathname: '/', search: '', hash: '' }),
          calcLocation: () => ({ name: 'state', pathMaskParams: [], availableParams: [], searchParams: [], pathMask: '/' })
        }
      }}>
        <Slot name="some"><span>Hello!</span></Slot>
      </WithContext>
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render fallback function', () => {
    const component = renderer.create(
      <WithContext context={{
        store: store({ routerState: { name: '', params: {} } }),
        router: {
          moveTo: () => {},
          onStateChange: () => {},
          getState: () => ({ name: '', params: {}, composition: {} }),
          createHref: () => ({ pathname: '/', search: '', hash: '' }),
          calcLocation: () => ({ name: 'state', pathMaskParams: [], availableParams: [], searchParams: [], pathMask: '/' })
        }
      }}>
        <Slot name="some">{() => (<span>Hello!</span>)}</Slot>
      </WithContext>
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render function', () => {
    const component = renderer.create(
      <WithContext context={{
        store: store({ routerState: { name: '', params: {} } }),
        router: {
          moveTo: () => {},
          onStateChange: () => {},
          getState: () => ({ name: '', params: {}, composition: { some: true } }),
          createHref: () => ({ pathname: '/', search: '', hash: '' }),
          calcLocation: () => ({ name: 'state', pathMaskParams: [], availableParams: [], searchParams: [], pathMask: '/' })
        }
      }}>
        <Slot name="some" render={() => (<span>Hello!</span>)} />
      </WithContext>
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render slot', () => {
    const component = renderer.create(
      <WithContext context={{
        store: store({ routerState: { name: '', params: {} } }),
        router: {
          moveTo: () => {},
          onStateChange: () => {},
          getState: () => ({ name: '', params: {}, composition: { some: () => <span>Hello</span> } }),
          createHref: () => ({ pathname: '/', search: '', hash: '' }),
          calcLocation: () => ({ name: 'state', pathMaskParams: [], availableParams: [], searchParams: [], pathMask: '/' })
        }
      }}>
        <Slot name="some" />
      </WithContext>
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
