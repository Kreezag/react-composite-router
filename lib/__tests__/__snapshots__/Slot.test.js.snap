// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should render fallback function 1`] = `
<span>
  Hello!
</span>
`;

exports[`rendering should render function 1`] = `
<span>
  Hello!
</span>
`;

exports[`rendering should render slot 1`] = `
<span>
  Hello
</span>
`;

exports[`rendering should render something 1`] = `
<span>
  Hello!
</span>
`;
